### PR TITLE
Memoizer docs

### PR DESCRIPTION
**Pull request by sbesson**
This Pull Request adds tests and documentation for the memoization functionalities of Bio-Formats

Using MATLAB as an example consumer where this memoization has been tested, the
description is amended to:
- describe a minimal example of usage of Memoizer
- explain Bio-Formats is not thread safe
- provide a example of Memoizer usage in a multi-threaded environment

The TestMemoizer tests are expanded to cover the examples described in the documentation, including its usage in multi-threaded mode.

To test this PR:
- verify the CI job running the MATLAB tests remains green
- review the MATLAB section of the developer documentation for Bio-Formats

/cc @emilroz @hflynn 

--no-rebase

### CODE REVIEW

File: components/formats-gpl/test/matlab/TestMemoizer.m [hunk 1]
```diff
@@ -35,28 +35,86 @@
...
             assertTrue(self.reader.isSavedToMemo());
             self.reader.close();
+        end
+        
+        function testReinitialization(self)
+            % Create Memoizer reader
+            self.reader = loci.formats.Memoizer(bfGetReader(), 0);
+            self.reader.setId(self.filepath);
+            
+            % Re-initialize the existingreader
```
Comment (mtbc) [inline components/formats-gpl/test/matlab/TestMemoizer.m]: missing space

File: docs/sphinx/developers/matlab-dev.txt [hunk 2]
```diff
@@ -284,3 +284,83 @@ and pass this object directly to ``bfsave``::
 
 For more information about the methods to store the metadata, see the
 :javadoc:`MetadataStore <loci/formats/meta/MetadataStore.html>` Javadoc page.
+
+.. _reader_performance:
+
+Improving reading performance
+-----------------------------
+
+Initializing a Bio-Formats reader can consume time and memory. Most of the
```
Comment (melissalinkert) [inline docs/sphinx/developers/matlab-dev.txt]: Maybe "substantial time and memory"?

File: docs/sphinx/developers/matlab-dev.txt [hunk 3]
```diff
@@ -284,3 +284,83 @@ and pass this object directly to ``bfsave``::
...
+initialization time is spend in the :javadoc:`setId(java.lang.String) <loci/formats/IFormatHandler.html#setId(java.lang.String)>` call. Various
+factors can impact the performance of this step including the file size, the
+amount of metadata in the image and also the file format itself.
+
+To decrease initialization time, a solution is to use the
+:javadoc:`loci.formats.Memoizer <loci/formats/Memoizer.html>` reader wrapper.
```
Comment (melissalinkert) [inline docs/sphinx/developers/matlab-dev.txt]: Possibly emphasize that the speed up is only after the first initialization?

File: docs/sphinx/developers/matlab-dev.txt [hunk 4]
```diff
@@ -284,3 +284,83 @@ and pass this object directly to ``bfsave``::
...
+``setId()`` with a reader decorated by the ``Memoizer`` on the same input file
+will load the reader from the memo file instead of performing a full reader
+initialization.
+
+More constructors are described in the
+:javadoc:`Memoizer javadocs <loci/formats/Memoizer.html>` allowing to control
+the minimal initialization time required before caching the reader and/or to
+define a root directory under which the reader should be cached.
+
+As Bio-Formats is not thread-safe, reader memoization offers new solution to
```
Comment (melissalinkert) [inline docs/sphinx/developers/matlab-dev.txt]: offers a new solution

File: docs/sphinx/developers/matlab-dev.txt [hunk 5]
```diff
@@ -284,3 +284,83 @@ and pass this object directly to ``bfsave``::
...
+increase reading performance when doing parallel work. For instance, the
+following example should how to combine memoization and MATLAB parfor to do
```
Comment (melissalinkert) [inline docs/sphinx/developers/matlab-dev.txt]: shows how

**Comment by melissalinkert** [comment]
Few minor nitpicks, but otherwise looks great.

**Comment by sbesson** [comment]
Pushed

**Comment by joshmoore** [comment]
Haven't read through, but a more general comment: the class-level docs in https://ci.openmicroscopy.org/job/BIOFORMATS-5.1-merge-build/javadoc/loci/formats/Memoizer.html look fairly sad. There might be something to be said for migrating some of the text here back to the Java file.

**Comment by melissalinkert** [comment]
Reads fine now - https://trello.com/c/0QWCVtOY/157-update-memoizer-javadocs created as a reminder to fix the Javadocs (possibly post-5.1.0).

**Pull request merged**